File: apps/android/gym/src/main/kotlin/works/windmill/gym/domain/Proposal.kt
package works.windmill.gym.domain

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

// Every type here is read-only on the wire: nothing on this phone mints a proposal, and one does
// nothing until the lifter taps Apply.

@Serializable(with = ProposalIntentSerializer::class)
enum class ProposalIntent(val wire: String) {
    Revise("revise"), Remove("remove");

    companion object {
        fun parse(raw: String?): ProposalIntent = entries.firstOrNull { it.wire == raw } ?: Revise
    }
}

object ProposalIntentSerializer : KSerializer<ProposalIntent> {
    override val descriptor = PrimitiveSerialDescriptor("ProposalIntent", PrimitiveKind.STRING)
    override fun serialize(encoder: Encoder, value: ProposalIntent) = encoder.encodeString(value.wire)
    override fun deserialize(decoder: Decoder): ProposalIntent = ProposalIntent.parse(decoder.decodeString())
}

// An unknown word reads as `Superseded`, never as `Pending`; an absent state takes `Pending`.
@Serializable(with = ProposalStateSerializer::class)
enum class ProposalState(val wire: String) {
    Pending("pending"), Applied("applied"), Dismissed("dismissed"), Superseded("superseded");

    companion object {
        fun parse(raw: String?): ProposalState = entries.firstOrNull { it.wire == raw } ?: Superseded
    }
}

object ProposalStateSerializer : KSerializer<ProposalState> {
    override val descriptor = PrimitiveSerialDescriptor("ProposalState", PrimitiveKind.STRING)
    override fun serialize(encoder: Encoder, value: ProposalState) = encoder.encodeString(value.wire)
    override fun deserialize(decoder: Decoder): ProposalState = ProposalState.parse(decoder.decodeString())
}

// `door` stays a String where `state` and `kind` are enums; `thread` is absent where there is no
// conversation to open.
@Serializable
data class ProposalSource(
    val door: String = "mcp",
    val connection: String? = null,
    val agent: String? = null,
    val thread: String? = null,
) {
    val conversation: String? get() = thread?.takeIf { it.isNotBlank() && door == askDoor }

    val name: String
        get() = named ?: if (door == askDoor) "Coach" else "your connected agent"

    // The prose is attributed by who wrote it: the agent by name, Coach for the room's own door, and
    // an agent that did not name itself is still the lifter's own. The same three on every surface.
    val kicker: String
        get() = "${named ?: if (door == askDoor) "Coach" else "Your agent"} wrote:"

    private val named: String?
        get() = agent?.takeIf { it.isNotBlank() } ?: connection?.takeIf { it.isNotBlank() }

    companion object {
        const val askDoor = "ask"
    }
}

// The absences: no reps is `max`, no weight is last time, no rest is the global dial, no sets is an
// open line. Which SIDE is missing is `kind`.
@Serializable
data class ProposalTargets(
    val sets: Int? = null,
    val reps: Int? = null,
    val weightKg: Double? = null,
    val restSeconds: Int? = null,
)

@Serializable(with = ChangeKindSerializer::class)
enum class ChangeKind(val wire: String) {
    Kept("kept"), Added("added"), Removed("removed"), Retargeted("retargeted");

    companion object {
        // An unknown kind reads as `Retargeted`, never as `Kept`, so no row is hidden from the count.
        fun parse(raw: String?): ChangeKind = entries.firstOrNull { it.wire == raw } ?: Retargeted
    }
}

object ChangeKindSerializer : KSerializer<ChangeKind> {
    override val descriptor = PrimitiveSerialDescriptor("ChangeKind", PrimitiveKind.STRING)
    override fun serialize(encoder: Encoder, value: ChangeKind) = encoder.encodeString(value.wire)
    override fun deserialize(decoder: Decoder): ChangeKind = ChangeKind.parse(decoder.decodeString())
}

// `before` is absent on an added line, `after` on a removed one. `loggedSets` rides on a removed line
// only, and a zero there is a real answer.
@Serializable
data class ProposalChange(
    val position: Int = 0,
    val kind: ChangeKind = ChangeKind.Retargeted,
    val exerciseId: String,
    val before: ProposalTargets? = null,
    val after: ProposalTargets? = null,
    val loggedSets: Int? = null,
) {
    fun addedLine(follows: String?): String {
        val said = mutableListOf("added")
        after?.let { said += Proposal.asks(it) }
        said += follows?.let { "after $it" } ?: "first in the routine"
        return said.joinToString(" · ")
    }

    // Zero is a real answer and an absent count says neither.
    val removedLine: String
        get() {
            val kept = loggedSets ?: return "removed from the routine"
            if (kept == 0) return "removed from the routine · never logged"
            val sets = if (kept == 1) "1 logged set" else "$kept logged sets"
            return "removed from the routine · $sets kept"
        }

    val compactLine: String
        get() {
            if (kind == ChangeKind.Added) return after?.let { "+ added · ${Proposal.asks(it)}" } ?: "+ added"
            if (kind == ChangeKind.Removed) return "− $removedLine"
            // A retarget, and any kind this build cannot name: read off whichever side arrived.
            val moved = before?.let { was -> after?.let { now -> Proposal.moves(was, now) } }.orEmpty()
            if (moved.isNotEmpty()) return moved.joinToString(" · ") { "${it.before} → ${it.after}" }
            return (after ?: before)?.let { Proposal.asks(it) } ?: "no targets"
        }
}

// Two sizes on one type: a HEAD carries everything above `baseRevision` and no diff. `changeCount` is
// the server's own count of what Apply does.
@Serializable
data class Proposal(
    val id: String,
    val routineId: String,
    val intent: ProposalIntent = ProposalIntent.Revise,
    val state: ProposalState = ProposalState.Pending,
    val summary: String = "",
    val changeCount: Int = 0,
    @SerialName("createdAt") val createdAtMs: Long = 0,
    @SerialName("settledAt") val settledAtMs: Long? = null,
    val source: ProposalSource = ProposalSource(),
    // Absent on a head, which is what tells the two sizes apart. Nullable and never 0.
    val baseRevision: Int? = null,
    val baseName: String = "",
    val name: String = "",
    val changes: List<ProposalChange> = emptyList(),
) {
    val isPending: Boolean get() = state == ProposalState.Pending

    val drawn: List<ProposalChange> get() = changes.filterNot { it.kind == ChangeKind.Kept }

    // The rows ARE the document as well as the diff: a kept row keeps its place in the run the routine
    // takes on, and a run of them collapses to one counted row that expands where it stands.
    val document: List<DocumentRow>
        get() {
            val rows = mutableListOf<DocumentRow>()
            var kept = mutableListOf<ProposalChange>()
            fun closeRun() {
                if (kept.isEmpty()) return
                rows += DocumentRow.Unchanged(kept.toList())
                kept = mutableListOf()
            }
            changes.forEach { change ->
                if (change.kind == ChangeKind.Kept) kept += change
                else {
                    closeRun()
                    rows += DocumentRow.Changed(change)
                }
            }
            closeRun()
            return rows
        }

    // Derived from the SERVER's reply and never from the model's prose: the name the routine now
    // carries and the server's own count of what moved, or that the routine is gone.
    val receipt: String?
        get() = when (state) {
            ProposalState.Applied -> {
                if (intent == ProposalIntent.Remove) "Applied · $routineName · routine removed"
                else "Applied · ${name.ifBlank { baseName }.ifBlank { "this routine" }} · $counted"
            }
            ProposalState.Dismissed -> turnedDownReceipt
            else -> null
        }

    val kicker: String get() = source.kicker

    val renames: Boolean
        get() = name.isNotBlank() && baseName.isNotBlank() && name != baseName

    fun landsAfter(added: ProposalChange): String? {
        val run = changes.takeWhile { it.kind != ChangeKind.Removed }
        val at = run.indexOf(added)
        if (at <= 0) return null
        return run[at - 1].exerciseId
    }

    // Only a routine that has moved PAST the base counts. A head answers no — it carries no base.
    fun supersededBy(routine: Routine?): Boolean {
        val base = baseRevision ?: return false
        val held = routine ?: return false
        return held.revision > base
    }

    fun byline(nowMs: Long): String = "from ${source.name} · ${Readout.whenLogged(createdAtMs, nowMs)}"

    fun summaryLine(routineName: String): String {
        if (summary.isNotBlank()) return summary
        if (intent == ProposalIntent.Remove) return "A proposal to remove $routineName."
        return "$counted to $routineName."
    }

    val counted: String
        get() {
            if (intent == ProposalIntent.Remove) return "a removal"
            return if (changeCount == 1) "1 change" else "$changeCount changes"
        }

    // One affordance on a card, and it is Review; the card's counted line says how much.
    val reviewLabel: String get() = review

    fun cardLine(routineName: String, stillWaiting: Boolean): String {
        val waiting = if (stillWaiting) Proposal.stillWaiting else "waiting"
        return "$routineName · $counted · $waiting"
    }

    // The server's count, never the rows this build drew: apply is atomic against the document.
    val applyLabel: String
        get() = when {
            intent == ProposalIntent.Remove -> "Remove $routineName"
            changeCount == 1 -> "Apply"
            else -> "Apply all $changeCount"
        }

    val atomicLine: String
        get() {
            if (intent == ProposalIntent.Remove) {
                return "The routine goes and your logged sets stay. Nothing is removed until you tap."
            }
            if (changeCount <= 1) return "Nothing is applied until you tap."
            return "All ${Readout.spelled(changeCount)} or none. Nothing is applied until you tap."
        }

    fun historyLine(nowMs: Long): String {
        val on = Readout.shortDate(settledAtMs ?: createdAtMs, nowMs)
        return when (state) {
            ProposalState.Applied -> "$on · applied $counted from ${source.name}"
            ProposalState.Dismissed -> "$on · turned down $counted from ${source.name}"
            ProposalState.Superseded -> "$on · set aside $counted from ${source.name}"
            ProposalState.Pending -> "$on · $counted from ${source.name}, waiting"
        }
    }

    fun settledNote(nowMs: Long): String? {
        val at = settledAtMs ?: return null
        val on = "${Readout.briefDay(at, nowMs)} at ${Readout.time(at)}"
        return when (state) {
            ProposalState.Applied ->
                "Applied to $routineName $on. Kept on the routine as a dated record — the program’s history, not a toast that disappears."
            ProposalState.Dismissed ->
                "Turned down $on. Nothing changed, and it stays in the routine’s history as a record."
            ProposalState.Superseded ->
                "$routineName changed after this was written, so it was set aside $on. None of it was applied, and it stays in the routine’s history."
            ProposalState.Pending -> null
        }
    }

    val routineName: String get() = baseName.ifBlank { "this routine" }

    companion object {
        const val review = "Review"
        const val apply = "Apply"

        // Why Apply is shut, said on the screen and on the semantics channel, and it names the way
        // out. Read off `seen` ALONE: bound to the disabled predicate it would say this while the
        // apply request is already on the wire.
        const val applyHint = "Read the changes to the end to apply them."
        const val stillWaiting = "still waiting"
        const val turnedDownReceipt = "Turned down · nothing changed."

        // Turning a proposal down is settled for good — the wire has no path back — so it is confirmed,
        // in the same words on every surface.
        const val turnDownVerb = "Turn this down"
        const val turnDownAsk = "Turn this down?"
        const val turnDownBody = "Nothing changes, and it stays in the routine’s history as a record."
        const val turnDown = "Turn down"

        // The load is compared on the LADDER's grid and never on raw doubles.
        fun moves(before: ProposalTargets, after: ProposalTargets): List<FieldMove> {
            val moved = mutableListOf<FieldMove>()
            if (before.sets != after.sets || before.reps != after.reps) {
                moved += FieldMove("sets", countOf(before), countOf(after))
            }
            if (!sameLoad(before.weightKg, after.weightKg)) {
                moved += FieldMove("weight", loadOf(before.weightKg), loadOf(after.weightKg))
            }
            if (before.restSeconds != after.restSeconds) {
                moved += FieldMove("rest", restOf(before.restSeconds), restOf(after.restSeconds))
            }
            return moved
        }

        fun asks(targets: ProposalTargets): String =
            Readout.target(targets.sets, targets.reps, targets.weightKg)

        fun countOf(targets: ProposalTargets): String {
            val sets = targets.sets ?: return Readout.openTarget
            return "$sets × ${Readout.repTarget(targets.reps)}"
        }

        fun loadOf(weightKg: Double?): String =
            weightKg?.let { Readout.weight(it) } ?: "last time"

        fun restOf(seconds: Int?): String = seconds?.let { "${it}s" } ?: "the dial"

        private fun sameLoad(before: Double?, after: Double?): Boolean {
            if (before == null || after == null) return before == null && after == null
            return Ladder.round(before) == Ladder.round(after)
        }
    }
}

// The routine is absent when the intent was to remove it.
@Serializable
data class ProposalDecision(val proposal: Proposal, val routine: Routine? = null)

sealed interface DocumentRow {
    data class Changed(val change: ProposalChange) : DocumentRow
    data class Unchanged(val kept: List<ProposalChange>) : DocumentRow {
        val label: String
            get() = if (kept.size == 1) "and 1 line unchanged" else "and ${kept.size} lines unchanged"
    }
}

data class FieldMove(val label: String, val before: String, val after: String)

File: apps/ios/WindmillKit/Sources/WindmillGym/Proposal.swift
import Foundation

// The native statement of backend/products/gym/domain/Proposal.h; every type here is `Decodable` only — a proposal is
// never written from this device.
// The enums are closed with no lenient default: an unknown word must fail the read rather than drop a row from a diff.

public enum ProposalIntent: String, Decodable, Sendable {
    case revise
    case remove
}

// The wire calls the fourth `superseded`; on screen it is set aside.
public enum ProposalState: String, Decodable, Sendable {
    case pending
    case applied
    case dismissed
    case superseded

    public var word: String {
        switch self {
        case .pending: return "Pending"
        case .applied: return "Applied"
        case .dismissed: return "Turned down"
        case .superseded: return "Set aside"
        }
    }
}

// `door` stays a String where `state` and `kind` are closed enums: an unknown door only picks a fallback name.
public struct ProposalSource: Equatable, Decodable, Sendable {
    public let door: String
    public let connection: String?
    public let agent: String?
    // nil is a fact, not a gap: there was no conversation, or it was deleted. Never inferred from `door == "ask"`.
    public let thread: String?

    public init(door: String, connection: String? = nil, agent: String? = nil,
                thread: String? = nil) {
        self.door = door
        self.connection = connection
        self.agent = agent
        self.thread = thread
    }

    // A connection's name counts as the agent's name, for the byline and the kicker alike.
    public var named: String? {
        [agent, connection].compactMap { $0 }.first { !$0.isEmpty }
    }

    public var agentName: String {
        guard let named else { return door == "ask" ? "Coach" : "your connected agent" }
        return named
    }
}

// `GET /v1/gym/proposals` answers every proposal on the account, newest first.
public struct ProposalHead: Equatable, Decodable, Sendable, Identifiable {
    public let id: String
    public let routineId: String
    public let intent: ProposalIntent
    public let state: ProposalState
    public let summary: String
    public let changeCount: Int
    public let createdAtMs: Int64
    public let settledAtMs: Int64?
    public let source: ProposalSource

    public init(id: String, routineId: String, intent: ProposalIntent = .revise,
                state: ProposalState = .pending, summary: String = "", changeCount: Int = 0,
                createdAtMs: Int64, settledAtMs: Int64? = nil,
                source: ProposalSource = ProposalSource(door: "mcp")) {
        self.id = id
        self.routineId = routineId
        self.intent = intent
        self.state = state
        self.summary = summary
        self.changeCount = changeCount
        self.createdAtMs = createdAtMs
        self.settledAtMs = settledAtMs
        self.source = source
    }

    public var isPending: Bool { state == .pending }

    public func line(about routineName: String) -> String {
        guard summary.isEmpty else { return summary }
        guard intent == .revise else { return "A proposal to remove \(routineName)." }
        return "\(changes) to \(routineName)."
    }

    // A count: the ledger keeps one pending proposal per door, so two doors put two on one routine.
    public static func waitingLine(_ count: Int) -> String {
        count == 1 ? "1 proposal" : "\(count) proposals"
    }

    // The date and the History order come off the same instant: a settled proposal belongs to the day it settled.
    public var recordedAtMs: Int64 { settledAtMs ?? createdAtMs }

    public func historyLine(now: Int64) -> String {
        let when = Readout.when(recordedAtMs, now: now)
        let what = intent == .revise ? changes : "a removal"
        switch state {
        case .pending: return "\(when) · \(what) from \(source.agentName), waiting"
        case .applied: return "\(when) · applied \(what) from \(source.agentName)"
        case .dismissed: return "\(when) · turned down \(what) from \(source.agentName)"
        case .superseded: return "\(when) · set aside \(what) from \(source.agentName)"
        }
    }

    public var changes: String { Readout.changeCount(changeCount) }

    // The model's prose is attributed by its door, never bare: Coach for the ask door, the agent or the connection
    // by name over MCP, and `Your agent` when the MCP source carries neither.
    public var kicker: String {
        if source.door == "ask" { return "Coach wrote:" }
        return "\(source.named ?? "Your agent") wrote:"
    }

    enum CodingKeys: String, CodingKey {
        case id, routineId, intent, state, summary, changeCount, source
        case createdAtMs = "createdAt"
        case settledAtMs = "settledAt"
    }

    public init(from decoder: Decoder) throws {
        let fields = try decoder.container(keyedBy: CodingKeys.self)
        id = try fields.decode(String.self, forKey: .id)
        routineId = try fields.decode(String.self, forKey: .routineId)
        intent = try fields.decode(ProposalIntent.self, forKey: .intent)
        state = try fields.decode(ProposalState.self, forKey: .state)
        summary = try fields.decodeIfPresent(String.self, forKey: .summary) ?? ""
        changeCount = try fields.decodeIfPresent(Int.self, forKey: .changeCount) ?? 0
        createdAtMs = try fields.decode(Int64.self, forKey: .createdAtMs)
        settledAtMs = try fields.decodeIfPresent(Int64.self, forKey: .settledAtMs)
        source = try fields.decodeIfPresent(ProposalSource.self, forKey: .source)
            ?? ProposalSource(door: "")
    }
}

// Rows up to the first `removed` are the run the routine would take on, in order; the rest are what it would take away.
// `before` is absent on an added row, `after` on a removed one, and `loggedSets` rides only on a removed one, where 0 is real.
public struct ProposalChange: Equatable, Decodable, Sendable {
    public enum Kind: String, Decodable, Sendable {
        case kept
        case added
        case removed
        case retargeted
    }

    // Absent reps is `max`, absent weight is last time's, absent rest is the global target, and absent sets is the open row.
    // Which side of a diff is missing is `kind`, never an absent `sets`.
    public struct Targets: Equatable, Decodable, Sendable {
        public let sets: Int?
        public let reps: Int?
        public let weightKg: Double?
        public let restSeconds: Int?

        public init(sets: Int? = nil, reps: Int? = nil, weightKg: Double? = nil,
                    restSeconds: Int? = nil) {
            self.sets = sets
            self.reps = reps
            self.weightKg = weightKg
            self.restSeconds = restSeconds
        }
    }

    public struct Move: Equatable, Sendable {
        public let field: String
        public let before: String
        public let after: String
    }

    public let position: Int
    public let kind: Kind
    public let exerciseId: String
    public let before: Targets?
    public let after: Targets?
    public let loggedSets: Int?

    public init(position: Int, kind: Kind, exerciseId: String,
                before: Targets? = nil, after: Targets? = nil, loggedSets: Int? = nil) {
        self.position = position
        self.kind = kind
        self.exerciseId = exerciseId
        self.before = before
        self.after = after
        self.loggedSets = loggedSets
    }

    // Sets and reps are one move rather than two.
    public var moves: [Move] {
        guard let before, let after else { return [] }
        var moved: [Move] = []
        if before.sets != after.sets || before.reps != after.reps {
            // The load is its own row, so no weight is passed here.
            moved.append(Move(field: "sets",
                              before: Readout.target(sets: before.sets, reps: before.reps,
                                                     weightKg: nil),
                              after: Readout.target(sets: after.sets, reps: after.reps,
                                                    weightKg: nil)))
        }
        if before.weightKg != after.weightKg {
            moved.append(Move(field: "weight",
                              before: before.weightKg.map(Readout.weight) ?? "—",
                              after: after.weightKg.map(Readout.weight) ?? "—"))
        }
        if before.restSeconds != after.restSeconds {
            moved.append(Move(field: "rest",
                              before: before.restSeconds.map { Readout.clock(Int64($0) * 1000) } ?? "—",
                              after: after.restSeconds.map { Readout.clock(Int64($0) * 1000) } ?? "—"))
        }
        return moved
    }

    public func addedLine(after previous: String?) -> String {
        var said = ["added"]
        if let after { said.append(Readout.target(sets: after.sets, reps: after.reps, weightKg: after.weightKg)) }
        said.append(previous.map { "after \($0)" } ?? "first in the routine")
        return said.joined(separator: " · ")
    }

    // The count is what the removal does not touch, and zero is a real answer.
    public var removedLine: String {
        guard let loggedSets else { return "removed from the routine" }
        guard loggedSets > 0 else { return "removed from the routine · never logged" }
        let sets = loggedSets == 1 ? "1 logged set" : "\(loggedSets) logged sets"
        return "removed from the routine · \(sets) kept"
    }

    enum CodingKeys: String, CodingKey {
        case position, kind, exerciseId, before, after, loggedSets
    }
}

// The rows are the document as well as the diff: every entry in its position, a kept line among them, and a renamed
// routine first. A change is a row that is not `kept`; the count under the button is the server's.
public enum ProposalRow: Equatable, Sendable {
    case renamed(from: String, to: String)
    // `follows` is read only on an added row.
    case entry(ProposalChange, follows: String?)
    case kept(ProposalChange)

    public var isKept: Bool {
        if case .kept = self { return true }
        return false
    }
}

// What the sheet draws: a changed row at full weight, a run of kept rows folded to a count IN PLACE, so expanding
// never shows the document out of order.
public enum ProposalBlock: Equatable, Sendable, Identifiable {
    case row(Int, ProposalRow)
    case unchanged(Int, [ProposalChange])

    public var id: Int {
        switch self {
        case .row(let index, _), .unchanged(let index, _): return index
        }
    }
}

// `baseRevision` is the token apply is atomic against; this device never checks it.
public struct Proposal: Equatable, Decodable, Sendable, Identifiable {
    public let head: ProposalHead
    public let baseRevision: Int
    public let baseName: String
    public let name: String
    public let changes: [ProposalChange]

    public init(head: ProposalHead, baseRevision: Int, baseName: String,
                name: String, changes: [ProposalChange]) {
        self.head = head
        self.baseRevision = baseRevision
        self.baseName = baseName
        self.name = name
        self.changes = changes
    }

    public var id: String { head.id }
    public var routineId: String { head.routineId }
    public var intent: ProposalIntent { head.intent }
    public var state: ProposalState { head.state }

    public var rows: [ProposalRow] {
        var drawn: [ProposalRow] = []
        if name != baseName { drawn.append(.renamed(from: baseName, to: name)) }
        for change in changes {
            guard change.kind != .kept else {
                drawn.append(.kept(change))
                continue
            }
            let previous = change.position > 1
                ? changes.first { $0.position == change.position - 1 }?.exerciseId
                : nil
            drawn.append(.entry(change, follows: previous))
        }
        return drawn
    }

    public var blocks: [ProposalBlock] { Proposal.folded(rows) }

    public static func folded(_ rows: [ProposalRow]) -> [ProposalBlock] {
        var blocks: [ProposalBlock] = []
        for (index, row) in rows.enumerated() {
            guard case .kept(let change) = row else {
                blocks.append(.row(index, row))
                continue
            }
            if case .unchanged(let start, let run) = blocks.last {
                blocks[blocks.count - 1] = .unchanged(start, run + [change])
            } else {
                blocks.append(.unchanged(index, [change]))
            }
        }
        return blocks
    }

    public static func unchangedLabel(_ count: Int) -> String {
        count == 1 ? "and 1 line unchanged" : "and \(count) lines unchanged"
    }

    // The server's count, never the rows this build drew: apply is atomic against the whole document.
    // Turning down is settled for good, so it is asked before it runs; closing the screen decides nothing.
    public static let turnDown = "Turn this down"
    public static let turnDownTitle = "Turn this down?"
    public static let turnDownBody = "Nothing changes, and it stays in the routine’s history as a record."
    public static let turnDownConfirm = "Turn down"
    public static let turnDownKeep = "Keep it"

    // Closing the sheet decides nothing; the card then says so.
    public static let waiting = "waiting"
    public static let stillWaiting = "still waiting"
    public static let review = "Review"
    public static let close = "Close"
    public static let applyHint = "Scroll to the end to apply."

    // The receipt is the server's reply, never the model's prose; it is not stored and it does not pretend to be.
    // A revision names the routine as it now stands (`name`, which a rename moved); a removal names what went.
    public static func receipt(applied: Proposal) -> String {
        guard applied.intent == .revise else { return removalReceipt(of: applied.baseName) }
        let name = applied.name.isEmpty ? applied.baseName : applied.name
        return "Applied · \(name) · \(Readout.changeCount(applied.head.changeCount))"
    }

    // A removal's reply carries no routine: the log confirmed the routine is gone, and that is the receipt.
    public static func removalReceipt(of baseName: String) -> String {
        "Applied · \(baseName) · routine removed"
    }

    public static let turnedDownReceipt = "Turned down · nothing changed."

    public var applyLabel: String {
        guard intent == .revise else { return "Remove \(baseName)" }
        if head.changeCount == 1 { return "Apply" }
        return "Apply all \(head.changeCount)"
    }

    public var footnote: String {
        guard intent == .revise else {
            return "Removes the routine from your program · every logged set stays."
        }
        guard head.changeCount > 1 else { return "Nothing is applied until you tap." }
        return "All \(Proposal.spelled(head.changeCount)) or none. Nothing is applied until you tap."
    }

    public func settledNote(now: Int64) -> String? {
        guard let settledAtMs = head.settledAtMs else { return nil }
        let when = "\(Readout.when(settledAtMs, now: now)) at \(Readout.time(settledAtMs))"
        switch state {
        case .pending:
            return nil
        case .applied:
            return "Applied to \(baseName) \(when). Kept on the routine as a dated record — the program’s history, not a toast that disappears."
        case .dismissed:
            return "Turned down \(when). Nothing changed, and it stays in the routine’s history as a record."
        case .superseded:
            return "\(baseName) changed after this was written, so it was set aside \(when). None of it was applied, and it stays in the routine’s history."
        }
    }

    private static func spelled(_ count: Int) -> String {
        let words = ["two", "three", "four", "five", "six", "seven",
                     "eight", "nine", "ten", "eleven", "twelve"]
        guard count >= 2, count - 2 < words.count else { return String(count) }
        return words[count - 2]
    }

    enum CodingKeys: String, CodingKey {
        case baseRevision, baseName, name, changes
    }

    // The head's fields are on the same flat object, so both decode from one container.
    public init(from decoder: Decoder) throws {
        head = try ProposalHead(from: decoder)
        let fields = try decoder.container(keyedBy: CodingKeys.self)
        baseRevision = try fields.decodeIfPresent(Int.self, forKey: .baseRevision) ?? 0
        baseName = try fields.decodeIfPresent(String.self, forKey: .baseName) ?? ""
        name = try fields.decodeIfPresent(String.self, forKey: .name) ?? ""
        changes = try fields.decodeIfPresent([ProposalChange].self, forKey: .changes) ?? []
    }
}

// The routine is absent when the intent was to remove.
public struct AppliedProposal: Equatable, Decodable, Sendable {
    public let proposal: Proposal
    public let routine: Routine?

    public init(proposal: Proposal, routine: Routine? = nil) {
        self.proposal = proposal
        self.routine = routine
    }

    enum CodingKeys: String, CodingKey {
        case proposal, routine
    }
}
